synapse/rest/client/room.py

```diff
@@ -61,6 +61,20 @@
 logger = logging.getLogger(__name__)
 
 
+def has_3pid_invite_keys(content: JsonDict) -> bool:
+    # if the request has medium and address keys, we should treat it as a 3pid invite.
+    if all(key in content for key in ("medium", "address")):
+        if all(key in content for key in ("id_server", "id_access_token")):
+            return True
+        # if it does not have id_server or id_access_token, we should treat that as an error.
+        raise SynapseError(
+            400,
+            "`id_server` and `id_access_token` are required when doing 3pid invite",
+            Codes.MISSING_PARAM,
+        )
+    return False
```

Comment:
this function makes my head explode. Sometimes it returns true, sometimes it returns false, and sometimes it raises a SynapseError? This feels hugely overcomplicated.

I suggest you move the checks for `id_server` and `id_access_token` into `on_POST`. And use a separate function for `create_room`.

Reply:
Should it still raise the SynapseError for missing_param or should it be handled by the mechanisms provided implicitly?

Reply:
> Should it still raise the SynapseError for missing_param

I don't understand the question. What is "it" here?

Here are some things which seem obvious to me:
 * Synapse should return a 400 error when it receives a request with missing parameters.
 * The easiest way to return a 400 error is by raising a `SynapseError`.
 * The name `has_3pid_invite_keys` suggests it's just doing a boolean check, and returns `true` or `false`. If it also does validation, it needs a different name. 

There's no point in me telling you how to write every line of code: I could just write it myself. Please, have a think about it and consider what *you* think makes for an elegant solution.